### ISSUE TITLE
OSX can't make

### ISSUE THREAD

**Issue by octylFractal** [hidefromkgb/DPengine_old#1]
Not sure what's up, but the make files don't appear to work.

``` shell
$ make           
mkdir -p .obj/core/gif
mkdir -p .obj/core/ogl
gcc -m32 -mstackrealign -I../core -Wall -fPIC -fvisibility=hidden -O2  -c lib.c -o .obj/lib.o
lib.c:95:11: error: no member named 'draw' in 'struct _ENGD'
    engd->draw = rscm;
    ~~~~  ^
lib.c:117:12: error: unknown type name 'MENU'
id Submenu(MENU *menu, id base) {
           ^
lib.c:136:30: error: use of undeclared identifier 'MFL_CCHK'
            if (menu->flgs & MFL_CCHK) {
                             ^
lib.c:137:54: error: use of undeclared identifier 'MFL_RCHK'
                setOnStateImage_(item, (menu->flgs & MFL_RCHK & ~MFL_CCHK)?
                                                     ^
./mac.h:285:116: note: expanded from macro 'setOnStateImage_'
#define setOnStateImage_(inst, i)                                      objc_msgSend(inst, LoadedObjCSelectors[24], i)
                                                                                                                   ^
lib.c:137:66: error: use of undeclared identifier 'MFL_CCHK'
                setOnStateImage_(item, (menu->flgs & MFL_RCHK & ~MFL_CCHK)?
                                                                 ^
./mac.h:285:116: note: expanded from macro 'setOnStateImage_'
#define setOnStateImage_(inst, i)                                      objc_msgSend(inst, LoadedObjCSelectors[24], i)
                                                                                                                   ^
lib.c:139:47: error: use of undeclared identifier 'MFL_VCHK'
                setState_(item, (menu->flgs & MFL_VCHK)?
                                              ^
./mac.h:287:127: note: expanded from macro 'setState_'
#define setState_(inst, s)                                             objc_msgSend(inst, LoadedObjCSelectors[26], (unsigned)(s))
                                                                                                                              ^
lib.c:150:45: error: use of undeclared identifier 'MFL_GRAY'
            setEnabled_(item, (menu->flgs & MFL_GRAY)? false : true);
                                            ^
./mac.h:280:123: note: expanded from macro 'setEnabled_'
#define setEnabled_(inst, b)                                           objc_msgSend(inst, LoadedObjCSelectors[19], (bool)(b))
                                                                                                                          ^
lib.c:163:8: error: unknown type name 'MENU'
inline MENU *OSSpecificMenu(ENGD *engd) {
       ^
lib.c:169:28: error: unknown type name 'MENU'
void EngineOpenContextMenu(MENU *menu) {
                           ^
lib.c:287:5: error: use of undeclared identifier 'MENU'
    MENU *item = (typeof(item))tag(this);
    ^
lib.c:287:11: error: use of undeclared identifier 'item'; did you mean 'cItem'?
    MENU *item = (typeof(item))tag(this);
          ^~~~
          cItem
/System/Library/Frameworks/CoreServices.framework/Frameworks/AE.framework/Headers/AERegistry.h:65:3: note: 'cItem' declared here
  cItem                         = 'citm', /*  0x6369746d  */
  ^
lib.c:287:26: error: use of undeclared identifier 'item'
    MENU *item = (typeof(item))tag(this);
                         ^
lib.c:289:9: error: use of undeclared identifier 'item'
    if (item) {
        ^
lib.c:290:9: warning: implicit declaration of function 'ProcessMenuItem' is invalid in C99 [-Wimplicit-function-declaration]
        ProcessMenuItem(item);
        ^
lib.c:290:25: error: use of undeclared identifier 'item'
        ProcessMenuItem(item);
                        ^
lib.c:291:13: error: use of undeclared identifier 'item'
        if (item->flgs & MFL_CCHK)
            ^
lib.c:291:26: error: use of undeclared identifier 'MFL_CCHK'
        if (item->flgs & MFL_CCHK)
                         ^
lib.c:292:30: error: use of undeclared identifier 'item'
            setState_(this, (item->flgs & MFL_VCHK)? NSOnState : NSOffState);
                             ^
./mac.h:287:127: note: expanded from macro 'setState_'
#define setState_(inst, s)                                             objc_msgSend(inst, LoadedObjCSelectors[26], (unsigned)(s))
                                                                                                                              ^
lib.c:292:43: error: use of undeclared identifier 'MFL_VCHK'
            setState_(this, (item->flgs & MFL_VCHK)? NSOnState : NSOffState);
                                          ^
./mac.h:287:127: note: expanded from macro 'setState_'
#define setState_(inst, s)                                             objc_msgSend(inst, LoadedObjCSelectors[26], (unsigned)(s))
                                                                                                                              ^
lib.c:326:16: error: no member named 'draw' in 'struct _ENGD'
    if (!engd->draw)
         ~~~~  ^
fatal error: too many errors emitted, stopping now [-ferror-limit=]
1 warning and 20 errors generated.
make: *** [.obj/lib.o] Error 1
```

**Comment by hidefromkgb** [comment]
Oh my.
I am sorry for being sidetracked for so long. All work and no play makes Jack a dull boy…
DPE is currently stuck in the middle of some major architectural changes I couldn`t yet find the guts to port across all OSes, OS X being one of them. Hopefully, not for long.
As a temporary measure, you can still build DPE from a less recent (oh irony) commit: https://github.com/hidefromkgb/DPengine/tree/04334cc5de14af0f9ec3d5aca5b075df46c8cabb
P.S.: how come GitHub not inform me of a new issue by email?

**Comment by octylFractal** [comment]
Sweet. Thanks for doing this btw :+1:

**Comment by octylFractal** [comment]
That seems to segfault, so I'll continue waiting patiently. Here's logs if you want them: https://gist.github.com/kenzierocks/217cf585283e318767fc

**Comment by hidefromkgb** [comment]
Thank you for the logs. What version of OS X are you using?
I\`ve never had (or wanted) a Mac, so it\`s extremely problematic for me to properly code for OS X. The only suitable way of doing this is running some sort of ancient Hackintosh — currently 10.6 x86-32 and 10.8 x86-64 — under VirtualBox, which my machine can barely handle.
The situation gets even worse with me using pure C-99 and despising standard OS X development tools like XCode, and ObjC itself. Too groovy for my taste.
So, may I humbly ask you to continue testing DPE under Mac when I begin fixing it?

**Comment by octylFractal** [comment]
Yes. I'm on Yosemite or 10.10.5, 64-bit.

**Comment by octylFractal** [comment]
Any update on this?

**Comment by hidefromkgb** [comment]
Dammit, that was long.
Finally I\`ve refined the core GUI architecture and ported all OS-dependent parts to Windows and Linux. See the [ponychan DP thread](https://www.ponychan.net/fan/res/458.html#2340) for details.
[On track](https://github.com/hidefromkgb/DPengine/commit/bd492b5d33ea56a9ab2552a3c13e44c11424c53f) to doing the same thing for OS X. Will try my best not to ragequit again at the sight of my 10.8.6 Hackintosh VM running at ≤ 5 FPS.

P.S.: unfortunately, the OS X commit I linked you to won\`t quite make it work now. However, [there is a commit](https://github.com/hidefromkgb/DPengine/commit/52342632ea87b3de68b3e72dce0c17c126a89729) that almost surely will. Unfortunately, it\`s no more compatible with the new core API, so it`ll take me time to bridge it to the new core.

**Comment by hidefromkgb** [comment]
@kenzierocks, f2f4eda1 finally fixed this.
Further OS X-related commits are experimental, so I can`t guarantee that they work on fresh OS X versions. I will be updating this post as I test them.

**Comment by octylFractal** [comment]
Hi, I sort of forgot about this for a while. I'm going to take everything for a test run, and report back.

**Comment by octylFractal** [comment]
And already having an error. I guess `kCGFontIndexMax` may no longer be a constant in macOS Sierra?

```
$ cd macos
$ make
mkdir -p .obj/core/gif
mkdir -p .obj/core/ogl
gcc -m32 -mstackrealign -I../core -Wall -fPIC -fvisibility=hidden -O2  -c ../core/core.c -o .obj/core/core.o
/var/folders/gq/5bmczc1j07j5fhrbk0nbd_yc0000gn/T//ccecEvzo.s:4808:11: warning: section "__textcoal_nt" is deprecated
        .section __TEXT,__textcoal_nt,coalesced,pure_instructions
                 ^      ~~~~~~~~~~~~~
/var/folders/gq/5bmczc1j07j5fhrbk0nbd_yc0000gn/T//ccecEvzo.s:4808:11: note: change section name to "__text"
        .section __TEXT,__textcoal_nt,coalesced,pure_instructions
                 ^      ~~~~~~~~~~~~~
gcc -m32 -mstackrealign -I../core -Wall -fPIC -fvisibility=hidden -O2  -c ../core/gif/gifstd.c -o .obj/core/gif/gifstd.o
gcc -m32 -mstackrealign -I../core -Wall -fPIC -fvisibility=hidden -O2  -c ../core/ogl/common.c -o .obj/core/ogl/common.o
/var/folders/gq/5bmczc1j07j5fhrbk0nbd_yc0000gn/T//cc7fbqcb.s:2791:11: warning: section "__textcoal_nt" is deprecated
        .section __TEXT,__textcoal_nt,coalesced,pure_instructions
                 ^      ~~~~~~~~~~~~~
/var/folders/gq/5bmczc1j07j5fhrbk0nbd_yc0000gn/T//cc7fbqcb.s:2791:11: note: change section name to "__text"
        .section __TEXT,__textcoal_nt,coalesced,pure_instructions
                 ^      ~~~~~~~~~~~~~
gcc -m32 -mstackrealign -I../core -Wall -fPIC -fvisibility=hidden -O2  -c ../core/ogl/oglstd.c -o .obj/core/ogl/oglstd.o
/var/folders/gq/5bmczc1j07j5fhrbk0nbd_yc0000gn/T//cc0jVaYs.s:1951:11: warning: section "__textcoal_nt" is deprecated
        .section __TEXT,__textcoal_nt,coalesced,pure_instructions
                 ^      ~~~~~~~~~~~~~
/var/folders/gq/5bmczc1j07j5fhrbk0nbd_yc0000gn/T//cc0jVaYs.s:1951:11: note: change section name to "__text"
        .section __TEXT,__textcoal_nt,coalesced,pure_instructions
                 ^      ~~~~~~~~~~~~~
gcc -m32 -mstackrealign -I../core -Wall -fPIC -fvisibility=hidden -O2  -c lib.c -o .obj/lib.o
In file included from /usr/include/Availability.h:176:0,
                 from /usr/include/sys/time.h:69,
                 from lib.c:1:
/System/Library/Frameworks/CoreFoundation.framework/Headers/CFDateFormatter.h:53:34: error: expected ',' or '}' before '__attribute__'
     kCFISO8601DateFormatWithYear API_AVAILABLE(macosx(10.12), ios(10.0), watchos(3.0), tvos(10.0)) = (1UL << 0),
                                  ^
/System/Library/Frameworks/CoreFoundation.framework/Headers/CFDateFormatter.h:80:126: error: 'introduced' undeclared here (not in a function)
 CFDateFormatterRef CFDateFormatterCreateISO8601Formatter(CFAllocatorRef allocator, CFISO8601DateFormatOptions formatOptions) API_AVAILABLE(macosx(10.12), ios(10.0), watchos(3.0), tvos(10.0));
                                                                                                                              ^
/System/Library/Frameworks/CoreFoundation.framework/Headers/CFURL.h:777:39: error: 'deprecated' undeclared here (not in a function)
 const CFStringRef kCFURLLabelColorKey API_DEPRECATED("Use NSURLLabelColorKey", macosx(10.6, 10.12), ios(4.0, 10.0), watchos(2.0, 3.0), tvos(9.0, 10.0));
                                       ^
/System/Library/Frameworks/CoreFoundation.framework/Headers/CFURL.h:777:39: error: 'message' undeclared here (not in a function)
 const CFStringRef kCFURLLabelColorKey API_DEPRECATED("Use NSURLLabelColorKey", macosx(10.6, 10.12), ios(4.0, 10.0), watchos(2.0, 3.0), tvos(9.0, 10.0));
                                       ^
In file included from /System/Library/Frameworks/Security.framework/Headers/Security.h:81:0,
                 from /System/Library/Frameworks/CoreServices.framework/Frameworks/OSServices.framework/Headers/CSIdentity.h:43,
                 from /System/Library/Frameworks/CoreServices.framework/Frameworks/OSServices.framework/Headers/OSServices.h:27,
                 from /System/Library/Frameworks/CoreServices.framework/Frameworks/LaunchServices.framework/Headers/IconsCore.h:23,
                 from /System/Library/Frameworks/CoreServices.framework/Frameworks/LaunchServices.framework/Headers/LaunchServices.h:22,
                 from /System/Library/Frameworks/CoreServices.framework/Headers/CoreServices.h:39,
                 from /System/Library/Frameworks/ApplicationServices.framework/Headers/ApplicationServices.h:23,
                 from mac.h:7,
                 from lib.c:4:
/System/Library/Frameworks/Security.framework/Headers/Authorization.h:194:14: error: variably modified 'bytes' at file scope
  char bytes[kAuthorizationExternalFormLength];
              ^
In file included from /usr/include/Availability.h:176:0,
                 from /usr/include/sys/time.h:69,
                 from lib.c:1:
/System/Library/Frameworks/CoreServices.framework/Frameworks/LaunchServices.framework/Headers/LSInfoDeprecated.h:1242:34: error: expected ',' or '}' before '__attribute__'
   kLSHandlerOptionsDefault       __OSX_AVAILABLE_BUT_DEPRECATED_MSG(__MAC_10_4, __MAC_10_11, __IPHONE_4_0, __IPHONE_9_0, "Creator codes are deprecated on OS X.") = 0,
                                  ^
/System/Library/Frameworks/CoreServices.framework/Frameworks/LaunchServices.framework/Headers/LSOpenDeprecated.h:27:33: error: expected ',' or '}' before '__attribute__'
   kLSLaunchInhibitBGOnly        __OSX_AVAILABLE_BUT_DEPRECATED_MSG(__MAC_10_0, __MAC_10_12, __IPHONE_NA, __IPHONE_NA, "This option does nothing.") = 0x00000080, /* Does nothing.*/
                                 ^
In file included from /System/Library/Frameworks/CoreGraphics.framework/Headers/CGContext.h:18:0,
                 from /System/Library/Frameworks/CoreGraphics.framework/Headers/CGBitmapContext.h:9,
                 from /System/Library/Frameworks/CoreGraphics.framework/Headers/CoreGraphics.h:11,
                 from /System/Library/Frameworks/ApplicationServices.framework/Headers/ApplicationServices.h:35,
                 from mac.h:7,
                 from lib.c:4:
/System/Library/Frameworks/CoreGraphics.framework/Headers/CGFont.h:53:1: error: initializer element is not constant
 static const CGFontIndex kCGGlyphMax = kCGFontIndexMax;
 ^
In file included from lib.c:4:0:
mac.h: In function 'PassT1FV':
mac.h:66:5: warning: dereferencing type-punned pointer will break strict-aliasing rules [-Wstrict-aliasing]
     return __PassT1FV(what);
     ^
make: *** [.obj/lib.o] Error 1
```

**Comment by hidefromkgb** [comment]
Just… wow. Weird errors in system headers, tons of unintelligible warnings… Seems that GCC is broken on 10.12.
Can you please try changing `gcc` to `clang` in the first two lines of `Makefile` and run `make` again?
According to the DP thread, on 10.11 it compiles and runs just fine (but still has problems deallocating memory on exit).
And, BTW, it`s now better to use b2440d4.

**Comment by octylFractal** [comment]
Changing it to `clang` works great!

![vinylscratch.png](https://i.imgur.com/8Z3o4YZ.png)

**Comment by hidefromkgb** [comment]
Yaaaay!
Ok then, can I mark this as resolved by 3acb435?

**Comment by octylFractal** [comment]
I think so! ❤️
